Animations- support for specifying Time type with QueryFilter

Leinnan opened the pull request:
# Objective
Fixes #20716

## Solution

Adds API that will make it possible to use different Time types for various animation players based on the FilterQuery.

## Testing

- I updated the animated_transform sample to use the new api and the virtual time instead of regular Time resource.

Leinnan committed:
Naive implementation

alice-i-cecile commented on crates/bevy_animation/src/lib.rs:
```diff
@@ -1220,6 +1227,22 @@ pub fn animate_targets(
         });
 }
 
+/// Adds animation support to an app
```
This needs more docs on  how it's used and why it exists.

alice-i-cecile reviewed (changes requested):
Needs a lot more docs to clearly explain why you might want to do this and how it can be used. But I really like the core idea, and don't hate the implementation.

IMO a usage example for this is more warranted, rather than bloating the existing example. We have a few clear use cases in mind: fixed updates for fighting game stability, or pausing animations when the game pauses. We should at least hint at that directly in the learning material.

I also think this warrants a release note: the usage in fighting games via fixed update is extremely interesting.

Leinnan committed:
Merge branch 'main' into feature/animations-time-typed

Leinnan committed:
Update docs

Leinnan committed:
Update lib.rs

Leinnan committed:
New example

Leinnan committed:
Update example

Leinnan committed:
Update docs and metadata

Leinnan committed:
Update docs

Leinnan committed:
Create animations_multiple_time_resources.md

Leinnan commented:
@alice-i-cecile Feedback applied 🫡 . One thing- I will be out for the next week, so if there some further changes required to make it into repo anyone else can feel free to create PR out of it instead if they can make it before I come back from vacation. The main thing for me is having that functionality built into the engine, not getting credit for it 😃

Leinnan committed:
Update lib.rs

Leinnan committed:
Sort `get_registered_reflect_types` response. It makes the response more
consistent and reduce chance of diff with the same content but in
different order

Leinnan committed:
Revert "Sort `get_registered_reflect_types` response. It makes the response more"

This reverts commit 9523c69538f7b925db7cbb1ed525400222c17d46.

alice-i-cecile reviewed (commented):
Much improved! The docs need a copyediting pass, but I'll wait to see what our Animation experts think about the architecture before fixing that up for you :)

Leinnan commented on crates/bevy_animation/src/lib.rs:
```diff
@@ -1220,6 +1228,25 @@ pub fn animate_targets(
+/// Specifies query for animation systems.
+/// It allows to specify which Time resource to use based on the [`bevy_ecs::query::QueryFilter`].
+/// That can be especially useful when you want to use a custom time resource for subset of animations.
+/// By default, there is added a one version with Time<()> and Without [`DontUseDefaultAnimationTime`] component.
```
It was

Leinnan committed:
Merge branch 'main' into feature/animations-time-typed

alice-i-cecile commented on crates/bevy_animation/src/lib.rs:
```diff
@@ -1220,6 +1228,25 @@ pub fn animate_targets(
+/// Specifies query for animation systems.
+/// It allows to specify which Time resource to use based on the [`bevy_ecs::query::QueryFilter`].
+/// That can be especially useful when you want to use a custom time resource for subset of animations.
+/// By default, there is added a one version with Time<()> and Without [`DontUseDefaultAnimationTime`] component.
```
```suggestion
/// By default, one version is added using [`Time<()>`] and a [`Without`] [`DontUseDefaultAnimationTime`] filter.
```

alice-i-cecile commented on crates/bevy_animation/src/lib.rs:
```diff
@@ -1220,6 +1228,25 @@ pub fn animate_targets(
         });
 }
 
+/// Specifies query for animation systems.
```
```suggestion
/// Specifies the query used for animation systems.
```

alice-i-cecile commented on crates/bevy_animation/src/lib.rs:
```diff
@@ -980,13 +980,21 @@ fn trigger_untargeted_animation_events(
+/// For more details, see the [`specify_animation_system`] documentation.
+#[derive(Debug, Component, Reflect, Clone, Copy, Default)]
+#[reflect(Component)]
+pub struct DontUseDefaultAnimationTime;
```
Bikeshedding help please: I don't like this name very much. It's extremely long, and the missed apostrophe bothers me.

alice-i-cecile commented on crates/bevy_animation/src/lib.rs:
```diff
@@ -1220,6 +1228,25 @@ pub fn animate_targets(
+/// It allows to specify which Time resource to use based on the [`bevy_ecs::query::QueryFilter`].
+/// That can be especially useful when you want to use a custom time resource for subset of animations.
+/// By default, there is added a one version with Time<()> and Without [`DontUseDefaultAnimationTime`] component.
+pub fn specify_animation_system<T: Default, F: bevy_ecs::query::QueryFilter + 'static>(
```
I don't think "specify" is the right word here. I think this would be a lot clearer if this was an ordinary struct-based plugin called `TimeDependentAnimationPlugin` or something

alice-i-cecile reviewed (changes requested):
Okay, so I'm pretty sold on the utility and the core implementation of this, but the clarity still needs a lot of work. I've left some suggestions that I think would help fix that.

Leinnan committed:
Update crates/bevy_animation/src/lib.rs

Co-authored-by: Alice Cecile <alice.i.cecile@gmail.com>

Leinnan committed:
Update crates/bevy_animation/src/lib.rs

Co-authored-by: Alice Cecile <alice.i.cecile@gmail.com>

Leinnan committed:
Update crates/bevy_animation/src/lib.rs

Co-authored-by: Alice Cecile <alice.i.cecile@gmail.com>

Leinnan committed:
Merge branch 'main' into feature/animations-time-typed

Leinnan committed:
Apply PR feedback

Leinnan committed:
Merge branch 'main' into feature/animations-time-typed

cart committed:
Merge branch 'main' into feature/animations-time-typed
